components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/util/RuleBuilder.java

```diff
@@ -216,4 +238,91 @@ private void setValidationError(String message) {
         isError = true;
         errorMessage = message;
     }
+
+    private Value validateAndResolveValue(FieldDefinition fieldDefinition, Value value) {
+
+        String rawValue = value.getFieldValue();
+
+        if (!isValidValueType(fieldDefinition, value)) {
+            return value;
+        }
+
+        if (!isValidOptionsInputValue(fieldDefinition, rawValue)) {
+            return value;
+        }
+
+        try {
+            return resolveValue(fieldDefinition, rawValue);
+        } catch (RuleManagementClientException e) {
+            setValidationError(e.getMessage());
+            return value;
+        }
+    }
+
+    private Value resolveValue(FieldDefinition fieldDefinition,
+                               String rawValue) throws RuleManagementClientException {
+
+        org.wso2.carbon.identity.rule.metadata.model.Value.ValueType
+                fieldDefinitionValueType = fieldDefinition.getValue().getValueType();
+
+        switch (fieldDefinitionValueType) {
+            case STRING:
+                return new Value(Value.Type.STRING, rawValue);
+            case NUMBER:
+                return validateNumberValue(rawValue);
+            case BOOLEAN:
+                return validateBooleanValue(rawValue);
+            case REFERENCE:
+                return new Value(Value.Type.REFERENCE, rawValue);
+            default:
+                throw new RuleManagementClientException(
```

Comment:
Is it okay to throw this client error for RAW type ?

Reply:
RAW type doesn't come here at all as that is applicable to the value of the expression not the metadata of the field of the expression.

This switch is about the value types allowed by rule metadata in an expression. Based on the field, it validates the raw value coming in request against the type defined in the metadata.